### PR TITLE
Refactor directional and punctual light evaluation

### PR DESCRIPTION
**Pull request by EvgeniiG**
CAREFUL REVIEW OF CODE REQUIRED

Goals:

- support volumetric/surface light/shadow dimmers in a general and efficient way;
- do not evaluate code if it's known to return 0;
- share code across all materials;
- bugfix volumetrics and transmission.

Katana: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=bugfix_volumetrics&automation-tools_branch=add-platform-filter&unity_branch=2018.3%2Fstaging

### CODE REVIEW

**Comment by EvgeniiG** [comment]
Note that I don't really like
#define MATERIAL_INCLUDE_TRANSMISSION
It would be good to get rid of it.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/HDLightEditor.Styles.cs [hunk 1]
```diff
@@ -27,6 +27,7 @@ sealed class Styles
             public readonly GUIContent lightIntensity = new GUIContent("Intensity", "");
 
             public readonly GUIContent maxSmoothness = new GUIContent("Max Smoothness", "Very low cost way of faking spherical area lighting. This will modify the roughness of the material lit. This is useful when the specular highlight is too small or too sharp.");
+            public readonly GUIContent lightRadius = new GUIContent("Light Radius", "Controls the size of the sphere light. For punctual and spot lights, it only affects volumetric lighting.");
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/HDLightEditor.Styles.cs]: Why only volumetric, this is also useful for direct lighting?
Or call it: Volumetric light radius

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightDefinition.cs [hunk 2]
```diff
@@ -50,61 +50,79 @@ public enum GPUImageBasedLightingType
...
+
+        public Vector4 shadowMaskSelector;      // Used with ShadowMask feature
+
+        public int     cookieIndex;             // -1 if unused
+        public int     tileCookie;              // (TODO: use a bitfield)
+        public int     nonLightMappedOnly;      // Used with ShadowMask feature (TODO: use a bitfield)
+        public float   minRoughness;            // This is use to give a small "area" to punctual light, as if we have a light with a radius.
+
+        public Vector2 size;                    // Used by area (X = length or width, Y = height) and punctual lights (X = radius^2)
+        public float   rangeAttenuationScale;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/LightDefinition.cs]: range attenaution is access before shadow code, for better cache usage it should be move above

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightEvaluation.hlsl [hunk 3]
```diff
@@ -151,37 +167,40 @@ float4 EvaluateCookie_Punctual(LightLoopContext lightLoopContext, LightData ligh
...
+                                           light.angleScale, light.angleOffset);
 
     // TODO: sample the extinction from the density V-buffer.
-    float distVol = (lightData.lightType == GPULIGHTTYPE_PROJECTOR_BOX) ? distances.w : distances.x;
+    float distVol = (light.lightType == GPULIGHTTYPE_PROJECTOR_BOX) ? distances.w : distances.x;
     attenuation *= TransmittanceHomogeneousMedium(_GlobalExtinction, distVol);
 
     // Projector lights always have cookies, so we can perform clipping inside the if().
-    UNITY_BRANCH if (lightData.cookieIndex >= 0)
+    UNITY_BRANCH if (light.cookieIndex >= 0)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/LightEvaluation.hlsl]: Any reason to keep UNITY_BRANCH here and not for directional cookie?

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.hlsl [hunk 4]
```diff
@@ -68,30 +69,52 @@ void LightLoop( float3 V, PositionInputs posInput, PreLightData preLightData, BS
...
+        // Evaluate sun shadows.
+        if (_DirectionalShadowIndex >= 0)
         {
-            context.shadowValue = 1.0f;
+            DirectionalLightData light = _DirectionalLightDatas[_DirectionalShadowIndex];
+
+            // TODO: this will cause us to load from the normal buffer first. Does this cause a performance problem?
+            // Also, the light direction is not consistent with the sun disk highlight hack, which modifies the light vector.
+            float  NdotL = dot(bsdfData.normalWS, -light.forward);
+            float3 shadowBiasNormal = GetShadowNormalBias(bsdfData); // Misleading name
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.hlsl]: you mean GetShadowNormalBias should be call GetShadowBiasNormal ? Let's rename it :)

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricLighting.compute [hunk 5]
```diff
@@ -136,23 +139,34 @@ VoxelLighting EvaluateVoxelLighting(LightLoopContext context, uint featureFlags,
...
-                                                              _DirectionalLightDatas[_DirectionalShadowIndex].shadowIndex,
-                                                             -_DirectionalLightDatas[_DirectionalShadowIndex].forward,
-                                                              posInput.positionSS);
+        // Evaluate sun shadows.
+        if (_DirectionalShadowIndex >= 0)
+        {
+            DirectionalLightData light = _DirectionalLightDatas[_DirectionalShadowIndex];
+
+            light.contactShadowIndex = -1;
+            light.lightDimmer        = light.volumetricLightDimmer;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricLighting.compute]: Add a comment here to say that we use lightDimmer/shadowDimmer in the called code so we transfer the volumetric prpoerty to it for code sharing (better worded of course :))

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricLighting.compute]: + add comment like "light.contactShadowIndex = -1;" // disable shadow contact as they are not use with volumetric

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricLighting.compute [hunk 6]
```diff
@@ -262,31 +280,35 @@ VoxelLighting EvaluateVoxelLighting(LightLoopContext context, uint featureFlags,
                 if (sampleLight)
                 {
                     // We are unable to adequately sample features larger
-                    // than the half of the length of the integration interval
-                    // divided by the number of temporal samples (7).
+                    // than the half of the length of the integration interval.
                     // Therefore, we apply this hack to reduce flickering.
-                    float hackMinDistSq = Sq(dt * (0.5 / 7));
+                    // This clamp is quite aggressive, blurring the light to kill all aliasing.
+                    // The number 7 comes from the number of unique temporal samples.
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricLighting.compute]: >The number 7
We should use a define then

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightEvaluation.hlsl (hunk 3 above)
Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/LightEvaluation.hlsl]: IMHO we should nuke all branches which will be inserted by the compiler, anyway, as all they do is make the code harder to read.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.cs [hunk 7]
```diff
@@ -759,93 +759,111 @@ bool GetDominantLightWithShadows(AdditionalShadowData additionalShadowData, Visi
...
+            lightData.minRoughness = Mathf.Max((1.0f - additionalLightData.maxSmoothness) * (1.0f - additionalLightData.maxSmoothness));
+
+            lightData.shadowMaskSelector = Vector4.zero;
 
             if (IsBakedShadowMaskLight(lightComponent))
             {
-                directionalLightData.shadowMaskSelector[lightComponent.bakingOutput.occlusionMaskChannel] = 1.0f;
-                directionalLightData.nonLightmappedOnly = lightComponent.lightShadowCasterMode == LightShadowCasterMode.NonLightmappedOnly ? 1 : 0;
+                lightData.shadowMaskSelector[lightComponent.bakingOutput.occlusionMaskChannel] = 1.0f;
+                // TODO: make this option per light, not global
```
[the remark this answers is not in the document]
Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.cs]: Done.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/HDLightEditor.Styles.cs (hunk 1 above)
Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/HDLightEditor.Styles.cs]: Done.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.cs [hunk 8]
```diff
@@ -999,22 +1025,20 @@ float ComputeLinearDistanceFade(float distanceToCamera, float fadeDistance)
                 lightData.angleOffset = 1.0f;
             }
 
+            if (lightData.lightType != GPULightType.Directional && lightData.lightType != GPULightType.ProjectorBox)
+            {
+                lightData.size = new Vector2(additionalLightData.shapeRadius * additionalLightData.shapeRadius, 0);
```
[the remark this answers is not in the document]
Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.cs]: Done.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.cs [hunk 9]
```diff
@@ -1061,13 +1087,13 @@ float ComputeLinearDistanceFade(float distanceToCamera, float fadeDistance)
             {
                 lightData.shadowMaskSelector[lightComponent.bakingOutput.occlusionMaskChannel] = 1.0f;
                 // TODO: make this option per light, not global
```
[the remark this answers is not in the document]
Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.cs]: Done.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.hlsl (hunk 4 above)
Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoop.hlsl]: Yes. Done.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricLighting.compute (hunk 6 above)
Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricLighting.compute]: This hack's been replaced by the fill light hack.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricLighting.compute (hunk 5 above)
Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricLighting.compute]: Done.

**Comment by EvgeniiG** [comment]
Feedback addressed, running Katana again.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/SurfaceShading.hlsl [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+
+// 15 degrees
+#define TRANSMISSION_WRAP_ANGLE (PI/12)
+#define TRANSMISSION_WRAP_LIGHT cos(PI/2 - TRANSMISSION_WRAP_ANGLE)
+
+//-----------------------------------------------------------------------------
+// Directional lights
+//-----------------------------------------------------------------------------
+
+float3 ComputeSunLightDirection(DirectionalLightData lightData, float3 N, float3 V)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/SurfaceShading.hlsl]: ComputeSunLightDirection should be in lightEvaluation.hlsl

Comment (EvgeniiG) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/SurfaceShading.hlsl]: It's not used for volumetrics.

**Pull request merged**